kratos/utilities/particles_utilities.h

```diff
@@ -0,0 +1,330 @@
+/** Detail class definition.
+*/
+class ParticlesUtilities
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of ParticlesUtilities
+    KRATOS_CLASS_POINTER_DEFINITION(ParticlesUtilities);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    ParticlesUtilities(){};
+
+    /// Destructor.
+    virtual ~ParticlesUtilities(){};
+
+    template<unsigned int TDim>
+    static void CountParticlesInNodes(
+            BinBasedFastPointLocator<TDim>& rLocator,
+            ModelPart& rVolumeModelPart,
+            const ModelPart& rParticlesModelPart,
+            const Variable<double>& rCounterVariable
+            )
+        {
+
+            //reset the EMBEDDED_VELOCITY
+            block_for_each(rVolumeModelPart.Nodes(), [&rCounterVariable](auto& rNode){
+                rNode.GetValue(rCounterVariable) = 0.0;
+            });
+
+
+            unsigned int max_results = 10000; 
+            typename BinBasedFastPointLocator<TDim>::ResultContainerType TLS(max_results);
+            
+            //for every interface node (nodes in cut elements)
+            block_for_each(rParticlesModelPart.Nodes(), TLS, [&rLocator, &rCounterVariable](const auto& rNode, auto& rTLS){
+
+                Vector shape_functions;
+                Element::Pointer p_element;
+                bool is_found = rLocator.FindPointOnMesh(rNode.Coordinates(), shape_functions, p_element, rTLS.begin(), rTLS.size(), 1e-5);
+
+                if(is_found){
+
+                    auto& r_geom = p_element->GetGeometry();
+                    for(unsigned int i=0; i<r_geom.size(); ++i){
+                        auto& rcounter = r_geom[i].GetValue(rCounterVariable);
+                        AtomicAdd(rcounter, 1.0);
+                    }
+                }
+
+            });
+
+        }
+
+    /**
+    this function looks if particle is found in the locator. If it is not, rVariable is marked with the value "outsider_value"
+    */
+    template<unsigned int TDim, class TScalarType>
+    static void MarkOutsiders(
+            BinBasedFastPointLocator<TDim>& rLocator,
+            ModelPart& rParticlesModelPart,
+            const Variable<TScalarType>& rVariable,
+            const TScalarType outsider_value
+            )
+        {
+            unsigned int max_results = 10000; 
+            typename BinBasedFastPointLocator<TDim>::ResultContainerType TLS(max_results);
+            
+            //for every interface node (nodes in cut elements)
+            block_for_each(rParticlesModelPart.Nodes(), TLS, [&rLocator, &rVariable, outsider_value](auto& rNode, auto& rTLS){
+
+                Vector shape_functions;
+                Element::Pointer p_element;
+                bool is_found = rLocator.FindPointOnMesh(rNode.Coordinates(), shape_functions, p_element, rTLS.begin(), rTLS.size(), 1e-5);
+
+                if(!is_found){
+                    rNode.SetValue(rVariable, outsider_value);
```

Comment:
Shouldn't you use a flag for this purpose? Besides, this would make possible to get rid of the `TScalarType` template argument.

Reply:
no, here it has to be a variable. The point is that i am assigning a "color" to the particles, and here i assign a special color to the ones that become "lost" to the simulation.

A more correct discussion however is that we should be able to set this either in the Database with history or in the one without history. I need to look how this was done e.g. in the variable utils

Reply:
Understood. In this case we need it to be a variable. About the `VariablesUtils` comment, it depends on the method you need. Most of them work for both historical and non-historical database, so I'd try to stick to the non-historical one as much as possible.